dot/sync/message.go

```diff
@@ -4,15 +4,48 @@
 package sync
 
 import (
+	"bytes"
 	"fmt"
 
 	"github.com/ChainSafe/gossamer/dot/network"
+	"github.com/ChainSafe/gossamer/dot/peerset"
 	"github.com/ChainSafe/gossamer/dot/types"
 	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/libp2p/go-libp2p/core/peer"
 )
 
+const maxNumberOfSameRequestPerPeer uint = 2
+
 // CreateBlockResponse creates a block response message from a block request message
-func (s *Service) CreateBlockResponse(req *network.BlockRequestMessage) (*network.BlockResponseMessage, error) {
+func (s *Service) CreateBlockResponse(from peer.ID, req *network.BlockRequestMessage) (
+	*network.BlockResponseMessage, error) {
+	logger.Debugf("sync request from %s: %s", from, req.String())
+
+	encodedRequest, err := req.Encode()
+	if err != nil {
+		return nil, fmt.Errorf("encoding request: %w", err)
+	}
+
+	encodedKey := bytes.Join([][]byte{[]byte(from.String()), encodedRequest}, nil)
+	requestHash, err := common.Blake2bHash(encodedKey)
+	if err != nil {
+		return nil, fmt.Errorf("hashing encoded block request sync message: %w", err)
+	}
+
+	numOfRequests := s.seenBlockSyncRequests.Get(requestHash)
+
+	if numOfRequests > maxNumberOfSameRequestPerPeer {
+		s.network.ReportPeer(peerset.ReputationChange{
+			Value:  peerset.SameBlockSyncRequest,
```

Comment:
Is this actually something to report to the network or should it just be internal?

Reply:
AFAIK there is no method to tell other peers about a spammer, so this should be manager internally. But I would assume that if the majority of the network implements the same anti-spam policy (same reputation value and threshold) then this bad actor will be rejected by them without much effort.

Reply:
Yeah that makes sense to me, however is this a defined "reason" for reporting that other nodes would have? Would this be handled only by gossamer nodes?

Reply:
This reason is a Gossamer specific field, when reporting a peer we also provide a reason for log purposes

Reply:
ohhhh gotcha, i was misunderstanding, thank ya